src/AutoBalance.cpp

```diff
@@ -832,59 +1299,108 @@ class AutoBalance_PlayerScript : public PlayerScript
             }
         }
 
-        virtual void OnLevelChanged(Player* player, uint8 /*oldlevel*/) override
+        virtual void OnLevelChanged(Player* player, uint8 oldlevel) override
         {
-            if (!EnableGlobal || !player)
-                return;
-
-            if (LevelScaling == 0)
+            LOG_DEBUG("module.AutoBalance", "AutoBalance_PlayerScript::OnLevelChanged(): {} has leveled from {} to {}", player->GetName(), oldlevel, player->getLevel());
+            if (!player || player->IsGameMaster())
                 return;
 
-            AutoBalanceMapInfo *mapABInfo=player->GetMap()->CustomData.GetDefault<AutoBalanceMapInfo>("AutoBalanceMapInfo");
+            Map* map = player->GetMap();
 
-            if (mapABInfo->mapLevel < player->getLevel())
-                mapABInfo->mapLevel = player->getLevel();
+            if (!map || !map->IsDungeon())
+                return;
 
+            // first update the map's player stats
+            UpdateMapPlayerStats(map);
 
+            // schedule all creatures for an update
+            lastConfigTime = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
         }
 
         void OnGiveXP(Player* player, uint32& amount, Unit* victim, uint8 /*xpSource*/) override
         {
             Map* map = player->GetMap();
+
+            // If this isn't a dungeon, make no changes
+            if (!map->IsDungeon() || !victim)
+                return;
+
             AutoBalanceMapInfo *mapABInfo=map->CustomData.GetDefault<AutoBalanceMapInfo>("AutoBalanceMapInfo");
 
-            if (victim && DungeonScaleDownXP && mapABInfo->enabled)
+            if (victim && RewardScalingXP && mapABInfo->enabled)
             {
                 Map* map = player->GetMap();
 
+                AutoBalanceCreatureInfo *creatureABInfo=victim->CustomData.GetDefault<AutoBalanceCreatureInfo>("AutoBalanceCreatureInfo");
+
                 if (map->IsDungeon())
                 {
-                    // Ensure that the players always get the same XP, even when entering the dungeon alone
-                    auto maxPlayerCount = ((InstanceMap*)sMapMgr->FindMap(map->GetId(), map->GetInstanceId()))->GetMaxPlayers();
-                    auto currentPlayerCount = map->GetPlayersCountExceptGMs();
-                    amount *= (float)currentPlayerCount / maxPlayerCount;
+                    if (RewardScalingMethod == "dynamic")
```

Comment:
minor: you could replace these strings `dynamic, fixed` into enum

Reply:
Done!

Reply:
nice!